### PR TITLE
spades: 3.15.5 -> 4.0.0

### PR DESCRIPTION
**Pull request by bzizou**
## Description of changes

Upgrade to Spades 4.0.0

- compilation bug with gcc 13 has been fixed upstream
- copytree.patch suggested and accepted upstream
 
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`) ---> `./result/bin/spades.py --test`
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sp/spades/package.nix [hunk 1]
```diff
@@ -2,33 +2,36 @@
...
-    hash = "sha256-FVw2QNVx8uexmgUDHR/Q0ZvYLfeF04hw+5O9JBsSu/o=";
+    hash = "sha256-B8AusdnZD2EaxzvdMN3CQu1RsAyKN1cYnoqBN62M+4s=";
   };
   sourceRoot = "SPAdes-${finalAttrs.version}/src";
 
-  env.CXXFLAGS = toString [
-    # GCC 13: error: 'uint32_t' does not name a type
-    "-include cstdint"
-  ];
+  patches = [ ./copytree.patch ];
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:23]:
```suggestion
  patches = [ ./copytree.patch ];
```
What is the source of this patch?
If this patch exist somewhere else, use `fetchpatch` for it and include a comment with issue/PR that originated it.

File: pkgs/by-name/sp/spades/package.nix [hunk 2]
```diff
@@ -2,33 +2,36 @@
...
+
+  cmakeFlags = [ "-DZLIB_ENABLE_TESTS=OFF" ];
 
   nativeBuildInputs = [ cmake ];
 
   buildInputs = [
-    zlib
+    zlib-ng
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:30]:
```suggestion
    zlib-ng
```
`z` is last letter of alphabet.

File: pkgs/by-name/sp/spades/package.nix (hunk 1 above)
Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix:23]: I made this patch and submitted it upstream... https://github.com/ablab/spades/pull/1314

File: pkgs/by-name/sp/spades/package.nix (hunk 2 above)
Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix]: Do you mean that we should order the list? Well, it's done. I also reformatted long lines.

File: pkgs/by-name/sp/spades/package.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/by-name/sp/spades/package.nix:24]: Congrats. And thanks for fixing things in upstream. As this makes upkeep easier.
You can `fetchpatch` it then. We usually avoid directly adding patches to nixpkgs unless necessary.
Also, if upstream accepts your patch, then you may just move forward the version reference.

File: pkgs/by-name/sp/spades/package.nix (hunk 2 above)
Comment (superherointj) [reply pkgs/by-name/sp/spades/package.nix]: Yes, whenever possible (not Python, there order matters), it's good to order alphabetically. Looks nicer also.

**Comment by superherointj** [comment]
@bzizou I want to mark this as draft until you are ready for review.

Possibilities:
1) Use `fetchpatch` now. (pinned to commit)
2) Wait for upstream to catch up with patch.

File: pkgs/by-name/sp/spades/package.nix [hunk 3]
```diff
@@ -1,40 +1,54 @@
...
+    readline
+    zlib-ng
   ];
 
   doCheck = true;
 
   meta = {
-    description = "St. Petersburg genome assembler, a toolkit for assembling and analyzing sequencing data";
+    description =
+      "St. Petersburg genome assembler, a toolkit for assembling and analyzing sequencing data";
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:51]:
```suggestion
    description = "St. Petersburg genome assembler, a toolkit for assembling and analyzing sequencing data";
```
I think you may be using the wrong version of `nixfmt`. Use `nixfmt-rfc-style`.

https://search.nixos.org/packages?channel=unstable&show=nixfmt-rfc-style&from=0&size=50&sort=relevance&type=packages&query=nixfmt

File: pkgs/by-name/sp/spades/package.nix [hunk 4]
```diff
@@ -1,40 +1,54 @@
...
 stdenv.mkDerivation (finalAttrs: {
   pname = "spades";
-  version = "3.15.5";
+  version = "4.0.0";
 
   src = fetchurl {
-    url = "https://github.com/ablab/spades/releases/download/v${finalAttrs.version}/SPAdes-${finalAttrs.version}.tar.gz";
-    hash = "sha256-FVw2QNVx8uexmgUDHR/Q0ZvYLfeF04hw+5O9JBsSu/o=";
+    url =
+      "https://github.com/ablab/spades/releases/download/v${finalAttrs.version}/SPAdes-${finalAttrs.version}.tar.gz";
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:20]:
```suggestion
    url = "https://github.com/ablab/spades/releases/download/v${finalAttrs.version}/SPAdes-${finalAttrs.version}.tar.gz";
```
I think you may be using the wrong version of `nixfmt`. Use `nixfmt-rfc-style`.

https://search.nixos.org/packages?channel=unstable&show=nixfmt-rfc-style&from=0&size=50&sort=relevance&type=packages&query=nixfmt

File: pkgs/by-name/sp/spades/package.nix [hunk 5]
```diff
@@ -1,40 +1,54 @@
...
+    hash = "sha256-B8AusdnZD2EaxzvdMN3CQu1RsAyKN1cYnoqBN62M+4s=";
   };
   sourceRoot = "SPAdes-${finalAttrs.version}/src";
 
-  env.CXXFLAGS = toString [
-    # GCC 13: error: 'uint32_t' does not name a type
-    "-include cstdint"
+  patches = [
+    (fetchpatch {
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:26]:
```suggestion
  patches = [
    # https://github.com/ablab/spades/pull/1314
    (fetchpatch {
```
Add upstream PR/patch reference as comment.

File: pkgs/by-name/sp/spades/package.nix [hunk 6]
```diff
@@ -1,40 +1,54 @@
...
+      name = "copytree.patch";
+      url = "https://github.com/ablab/spades/pull/1314/commits/af9feaf4cc7cbd9102362309f582e8645c471e02.patch";
+      sha256 = "sha256-tkT7hb0TqsbLkcTs9u43nzvV8bVdh3G9VKYqFFLrQv8=";
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:29]:
```suggestion
      hash = "sha256-tkT7hb0TqsbLkcTs9u43nzvV8bVdh3G9VKYqFFLrQv8=";
```
Prefer `hash` over `sha256`.

File: pkgs/by-name/sp/spades/package.nix (hunk 3 above)
Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix:51]: Great! I was wondering why it did not format well...

**Comment by superherointj** [comment]
@bzizou PR seems ready for review? On yes, mark it as ready for review.

**Comment by superherointj** [comment]
Can you add:

> meta.changelog = "https://github.com/ablab/spades/blob/next/changelog.md";
> meta.downloadPage = "https://github.com/ablab/spades";

File: pkgs/by-name/sp/spades/package.nix [hunk 7]
```diff
@@ -1,40 +1,55 @@
...
     python3
+    readline
+    zlib-ng
   ];
 
   doCheck = true;
 
   meta = {
     description = "St. Petersburg genome assembler, a toolkit for assembling and analyzing sequencing data";
+    changelog = "https://github.com/ablab/spades/blob/next/changelog.md";
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:51]:
```suggestion
    changelog = "https://github.com/ablab/spades/blob/${finalAttrs.version}/changelog.md";
```
Sorry. Had to pin the version.

File: pkgs/by-name/sp/spades/package.nix [hunk 8]
```diff
@@ -1,42 +1,57 @@
 {
   lib,
   stdenv,
-  fetchurl,
-  zlib,
   bzip2,
   cmake,
+  fetchurl,
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:6]:
```suggestion
  fetchurl,
```
Is there a good reason for not using `fetchFromGitHub`? (Which is the preferable fetcher abstraction for GitHub)

Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix:6]: Checking...

Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix:6]: Sorry, had to rebuild as tarballs seems to differ... and tests are not passing anymore...  still checking why...

Comment (bzizou) [reply pkgs/by-name/sp/spades/package.nix:6]: The file https://github.com/ablab/spades/blob/next/src/CMakeListsInternal.txt is empty in the SPAdes-4.0.0.tar.gz release file. That makes the "github" version tests failing as they probably included some tests that are to be run only from their dev environement.
I thought that the flag `-DSPADES_BUILD_INTERNAL=OFF` would disable this, but it does nothing...
So, I emptied the CMakeListsInternal.txt as they did for their release file...

**Comment by superherointj** [comment]
Git commit message **body** (the second line after title) is missing the release notes:

> Release: https://github.com/ablab/spades/releases/tag/v4.0.0

**Review by superherointj — APPROVED** [review]

**Comment by superherointj** [comment]
This package is broken for `pkgsMusl`.

> nix build .#pkgsMusl.spades

Logs: https://termbin.com/sik1

Either add:
> broken = stdenv.hostPlatform.isMusl;

Or provide a fix for Musl.

**Comment by superherointj** [comment]
Enable strictDeps. I think, python may be needed at 'nativeBuildInputs'.

**Comment by superherointj** [comment]
@ofborg eval

Suspicion: `fetchpatch` of `copytree.patch` is failing.

**Comment by superherointj** [comment]
`fetchpatch` of `copytree.patch` is failing at `nixpkgs-review`

File: pkgs/by-name/sp/spades/package.nix [hunk 9]
```diff
@@ -1,42 +1,68 @@
...
-  env.CXXFLAGS = toString [
-    # GCC 13: error: 'uint32_t' does not name a type
-    "-include cstdint"
+  sourceRoot = "source/src";
+
+  patches = [
+    # https://github.com/ablab/spades/pull/1314
+    (fetchpatch {
+      name = "copytree.patch";
+      url = "https://github.com/ablab/spades/pull/1314/commits/af9feaf4cc7cbd9102362309f582e8645c471e02.patch";
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:31]:
```suggestion
      url = "https://github.com/ablab/spades/commit/af1f756a46c5da669897b841d4f753af1eaa9588.patch";
```

**Review by superherointj — DISMISSED** [review]
`fetchpatch` of `copytree.patch` is failing at `nixpkgs-review`

**Comment by bzizou** [comment]
I agree that redistributing dependencies is dirty...  they are numerous: https://github.com/ablab/spades/tree/next/ext/src

I tried to disable some, to replace them by dependencies to nix package but their Cmake files are not ok, and this leads to includes or libraries not found. My knowledge in cmake is probably not sufficient to find some solutions... and I already spent too much time on this. Sorry. So --> `broken = stdenv.hostPlatform.isMusl;` Maybe should I also add `broken = stdenv.hostPlatform.isStatic;` ?

**Comment by superherointj** [comment]
> I agree that redistributing dependencies is dirty... they are numerous: https://github.com/ablab/spades/tree/next/ext/src
> 
> I tried to disable some, to replace them by dependencies to nix package but their Cmake files are not ok, and this leads to includes or libraries not found. My knowledge in cmake is probably not sufficient to find some solutions... and I already spent too much time on this. Sorry. So --> `broken = stdenv.hostPlatform.isMusl;` Maybe should I also add `broken = stdenv.hostPlatform.isStatic;` ?

it is fine if you mark it broken.

File: pkgs/by-name/sp/spades/package.nix [hunk 10]
```diff
@@ -1,46 +1,74 @@
...
-    zlib
     bzip2
+    ncurses
     python3
+    readline
   ];
 
   doCheck = true;
 
+  strictDeps = true;
```
Comment (superherointj) [inline pkgs/by-name/sp/spades/package.nix:58]: > strictDeps = true;

strictDeps is configure phase.

**Comment by superherointj** [comment]
@bzizou LGTM. Only the Darwin build needs to be confirmed.
By any chance, are you able to build this package for Darwin?
If not, we can ping Darwin maintainers for this.

**Comment by superherointj** [comment]
@toonn Can you post build results for Darwin? Thanks. Sorry for the inconvenience. CI is just unreliable/slow for Darwin?

**Review by superherointj — APPROVED** [review]
I'm approving this despite Darwin build being unknown.

Update: Watching build at:
https://hydra.nixos.org/build/264908242

**Comment by bzizou** [comment]
> @bzizou LGTM. Only the Darwin build needs to be confirmed. By any chance, are you able to build this package for Darwin? If not, we can ping Darwin maintainers for this.

No, I've got no Darwin host

**Comment by superherointj** [comment]
Darwin build is broken:

https://hydra.nixos.org/build/264908242/nixlog/1

@bzizou Can it be fixed or should we mark it broken for Darwin?

**Pull request merged**